File: javascript/src/address.ts
/**
 * CiteKit Address Parser — parse and build URI-style addresses.
 *
 * Formats:
 *   doc://resource_id#pages=3-5
 *   video://resource_id#t=192-230
 *   audio://resource_id#t=60-120
 *   image://resource_id#bbox=0.2,0.3,0.8,0.7
 */

import type { Location } from "./models.js";

const SCHEME_TO_MODALITY: Record<string, Location["modality"]> = {
    doc: "document",
    video: "video",
    audio: "audio",
    image: "image",
    text: "text",
    virtual: "virtual",
};

const MODALITY_TO_SCHEME: Record<string, string> = {
    document: "doc",
    video: "video",
    audio: "audio",
    image: "image",
    text: "text",
    virtual: "virtual",
};

/**
 * Parse a CiteKit address into a resource ID and location.
 *
 * @example
 * parseAddress("doc://calculus_book#pages=12-13")
 * // → { resourceId: "calculus_book", location: { modality: "document", pages: [12, 13] } }
 */
export function parseAddress(address: string): {
    resourceId: string;
    location: Location;
} {
    const match = address.match(/^(\w+):\/\/([^#]+)(?:#(.+))?$/);
    if (!match) {
        throw new Error(`Invalid CiteKit address: ${address}`);
    }

    const [, scheme, resourceId, fragment] = match;

    const modality = SCHEME_TO_MODALITY[scheme];
    if (!modality) {
        throw new Error(
            `Unknown scheme '${scheme}'. Expected: ${Object.keys(SCHEME_TO_MODALITY).join(", ")}`
        );
    }

    const location: Location = { modality };

    if (scheme === "virtual") {
        location.virtual_address = address;
        return { resourceId, location };
    }

    if (fragment) {
        const params = Object.fromEntries(
            fragment.split("&").map((part) => {
                const [key, value] = part.split("=", 2);
                return [key, value];
            })
        );

        if (params.pages) {
            if (params.pages.includes("-")) {
                const [start, end] = params.pages.split("-").map(Number);
                location.pages = Array.from(
                    { length: end - start + 1 },
                    (_, i) => start + i
                );
            } else {
                location.pages = params.pages.split(",").map(Number);
            }
        }

        if (params.t) {
            const [start, end] = params.t.split("-").map(parseTime);
            location.start = start;
            location.end = end;
        }

        if (params.bbox) {
            const parts = params.bbox.split(",").map(Number);
            if (parts.length !== 4) {
                throw new Error(`bbox must have 4 values, got ${parts.length}`);
            }
            location.bbox = parts as [number, number, number, number];
        }

        if (params.lines) {
            const [start, end] = params.lines.split("-").map(Number);
            if (Number.isNaN(start) || Number.isNaN(end)) {
                throw new Error(`lines must be numeric, got '${params.lines}'`);
            }
            location.lines = [start, end];
        }
    }

    return { resourceId, location };
}

/**
 * Build a CiteKit address from a resource ID and location.
 *
 * @example
 * buildAddress("book", { modality: "document", pages: [3, 4, 5] })
 * // → "doc://book#pages=3-5"
 */
export function buildAddress(resourceId: string, location: Location): string {
    const scheme = MODALITY_TO_SCHEME[location.modality];
    if (!scheme) {
        throw new Error(`Unknown modality: ${location.modality}`);
    }

    if (location.modality === "virtual") {
        return location.virtual_address ?? `virtual://${resourceId}`;
    }

    const fragments: string[] = [];

    if (location.pages && location.pages.length > 0) {
        const sorted = [...location.pages].sort((a, b) => a - b);
        const isConsecutive =
            sorted.length === sorted[sorted.length - 1] - sorted[0] + 1;

        if (isConsecutive) {
            fragments.push(`pages=${sorted[0]}-${sorted[sorted.length - 1]}`);
        } else {
            fragments.push(`pages=${sorted.join(",")}`);
        }
    }

    if (location.start !== undefined && location.end !== undefined) {
        fragments.push(`t=${formatTime(location.start)}-${formatTime(location.end)}`);
    }

    if (location.bbox) {
        fragments.push(`bbox=${location.bbox.join(",")}`);
    }

    if (location.lines) {
        fragments.push(`lines=${location.lines[0]}-${location.lines[1]}`);
    }

    const fragment = fragments.join("&");
    return fragment ? `${scheme}://${resourceId}#${fragment}` : `${scheme}://${resourceId}`;
}

// ── Helpers ────────────────────────────────────────────────────────────────

function parseTime(timeStr: string): number {
    const trimmed = timeStr.trim();
    if (trimmed.includes(":")) {
        const parts = trimmed.split(":").map(Number);
        if (parts.length === 3) return parts[0] * 3600 + parts[1] * 60 + parts[2];
        if (parts.length === 2) return parts[0] * 60 + parts[1];
    }
    return Number(trimmed);
}

function formatTime(seconds: number): string {
    if (Number.isInteger(seconds)) {
        if (seconds >= 3600) {
            const h = Math.floor(seconds / 3600);
            const m = Math.floor((seconds % 3600) / 60);
            const s = seconds % 60;
            return `${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}:${String(s).padStart(2, "0")}`;
        }
        if (seconds >= 60) {
            const m = Math.floor(seconds / 60);
            const s = seconds % 60;
            return `${String(m).padStart(2, "0")}:${String(s).padStart(2, "0")}`;
        }
    }
    return String(seconds);
}

File: python/citekit/address.py
"""Address parser and builder for CiteKit URI-style addresses.

Formats:
    doc://resource_id#pages=3-5
    video://resource_id#t=192-230
    audio://resource_id#t=60-120
    image://resource_id#bbox=0.2,0.3,0.8,0.7
"""

from __future__ import annotations

import re
from urllib.parse import urlparse, parse_qs

from citekit.models import Location

# Scheme → modality mapping
_SCHEME_TO_MODALITY = {
    "doc": "document",
    "video": "video",
    "audio": "audio",
    "image": "image",
    "text": "text",
    "virtual": "virtual",
}

_MODALITY_TO_SCHEME = {v: k for k, v in _SCHEME_TO_MODALITY.items()}


def parse_address(address: str) -> tuple[str, Location]:
    """Parse a CiteKit address into (resource_id, Location).

    Examples:
        >>> parse_address("doc://calculus_book#pages=12-13")
        ('calculus_book', Location(modality='document', pages=[12, 13]))

        >>> parse_address("video://lecture1#t=192-230")
        ('lecture1', Location(modality='video', start=192.0, end=230.0))
        
        >>> parse_address("text://code#lines=10-20")
        ('code', Location(modality='text', lines=(10, 20)))

        >>> parse_address("image://diagram#bbox=0.2,0.3,0.8,0.7")
        ('diagram', Location(modality='image', bbox=(0.2, 0.3, 0.8, 0.7)))
    """
    # Parse scheme manually since urllib doesn't handle custom schemes well
    match = re.match(r"^(\w+)://([^#]+)(?:#(.+))?$", address)
    if not match:
        raise ValueError(f"Invalid CiteKit address: {address}")

    scheme, resource_id, fragment = match.group(1), match.group(2), match.group(3)

    if scheme not in _SCHEME_TO_MODALITY:
        raise ValueError(f"Unknown scheme '{scheme}'. Expected one of: {list(_SCHEME_TO_MODALITY.keys())}")

    modality = _SCHEME_TO_MODALITY[scheme]

    if scheme == "virtual":
        return resource_id, Location(
            modality="virtual",
            virtual_address=address,
        )

    # Parse fragment parameters
    pages = None
    start = None
    end = None
    bbox = None
    lines = None

    if fragment:
        params = dict(part.split("=", 1) for part in fragment.split("&") if "=" in part)

        if "pages" in params:
            page_str = params["pages"]
            if "-" in page_str:
                p_start, p_end = page_str.split("-", 1)
                pages = list(range(int(p_start), int(p_end) + 1))
            else:
                pages = [int(p) for p in page_str.split(",")]

        if "lines" in params:
            lines_str = params["lines"]
            if "-" in lines_str:
                l_start, l_end = lines_str.split("-", 1)
                lines = (int(l_start), int(l_end))
            elif "," in lines_str:
                # Fallback if someone uses comma, but lines is usually a range tuple
                parts = lines_str.split(",")
                if len(parts) >= 2:
                    lines = (int(parts[0]), int(parts[-1]))
                else:
                    lines = (int(parts[0]), int(parts[0]))

        if "t" in params:
            time_str = params["t"]
            if "-" in time_str:
                t_start, t_end = time_str.split("-", 1)
                start = _parse_time(t_start)
                end = _parse_time(t_end)

        if "bbox" in params:
            parts = params["bbox"].split(",")
            if len(parts) != 4:
                raise ValueError(f"bbox must have 4 values, got {len(parts)}")
            bbox = tuple(float(p) for p in parts)

    return resource_id, Location(
        modality=modality,
        pages=pages,
        start=start,
        end=end,
        bbox=bbox,
        lines=lines,
    )


def build_address(resource_id: str, location: Location) -> str:
    """Build a CiteKit address from a resource ID and location.

    Examples:
        >>> build_address("book", Location(modality="document", pages=[3, 4, 5]))
        'doc://book#pages=3-5'

        >>> build_address("lecture", Location(modality="video", start=192.0, end=230.0))
        'video://lecture#t=192-230'
    """
    scheme = _MODALITY_TO_SCHEME.get(location.modality)
    if not scheme:
        raise ValueError(f"Unknown modality: {location.modality}")

    if location.modality == "virtual":
        return location.virtual_address or f"virtual://{resource_id}"

    fragment_parts = []

    if location.pages is not None:
        if len(location.pages) == 0:
            raise ValueError("Pages list cannot be empty")
        pages_sorted = sorted(location.pages)
        # Check if pages are consecutive for range notation
        if pages_sorted == list(range(pages_sorted[0], pages_sorted[-1] + 1)):
            fragment_parts.append(f"pages={pages_sorted[0]}-{pages_sorted[-1]}")
        else:
            fragment_parts.append(f"pages={','.join(str(p) for p in pages_sorted)}")

    if location.lines is not None:
        start_line, end_line = location.lines
        fragment_parts.append(f"lines={start_line}-{end_line}")

    if location.start is not None and location.end is not None:
        start_str = _format_time(location.start)
        end_str = _format_time(location.end)
        fragment_parts.append(f"t={start_str}-{end_str}")

    if location.bbox is not None:
        bbox_str = ",".join(f"{v:g}" for v in location.bbox)
        fragment_parts.append(f"bbox={bbox_str}")

    fragment = "&".join(fragment_parts)
    if fragment:
        return f"{scheme}://{resource_id}#{fragment}"
    return f"{scheme}://{resource_id}"


def _parse_time(time_str: str) -> float:
    """Parse a time string to seconds. Supports seconds or HH:MM:SS format."""
    time_str = time_str.strip()

    # HH:MM:SS or MM:SS format
    if ":" in time_str:
        parts = time_str.split(":")
        if len(parts) == 3:
            h, m, s = parts
            return int(h) * 3600 + int(m) * 60 + float(s)
        elif len(parts) == 2:
            m, s = parts
            return int(m) * 60 + float(s)

    return float(time_str)


def _format_time(seconds: float) -> str:
    """Format seconds as a compact time string."""
    if seconds == int(seconds):
        seconds = int(seconds)

    # Use HH:MM:SS for large values
    if isinstance(seconds, int) and seconds >= 3600:
        h = seconds // 3600
        m = (seconds % 3600) // 60
        s = seconds % 60
        return f"{h:02d}:{m:02d}:{s:02d}"
    elif isinstance(seconds, int) and seconds >= 60:
        m = seconds // 60
        s = seconds % 60
        return f"{m:02d}:{s:02d}"

    return str(seconds)
